conans/test/functional/test_third_party_patch_flow.py

```diff
@@ -0,0 +1,199 @@
+    assert "conanfile.py (mypkg/1.0): Calling build()" in client.out
+
+    # of course create should work too
+    client.run("create .")
+    assert "mypkg/1.0: Created package" in client.out
+
+
+def test_third_party_overwrite_build_file():
+    """ this test emulates the work of a developer contributing recipes to ConanCenter, and
+    replacing the original build script with your one one.
+
+    The export_sources is actually copying CMakeLists.txt into the "src" folder, but the
+    'download' will overwrite it, so it is necessary to copy it again
+    """
+    conanfile = textwrap.dedent(r"""
+        import os, shutil
+        from conans import ConanFile, load
+        from conans.tools import save
+
+        class Pkg(ConanFile):
+            name = "mypkg"
+            version = "1.0"
+            exports_sources = "CMakeLists.txt"
+
+            def layout(self):
+                self.folders.source = "src"
+                self.folders.build = "build"
+
+            def source(self):
+                # emulate a download from web site
+                save("CMakeLists.txt", "MISTAKE: Very old CMakeLists that needs to be replaced")
+                # Now I fix it with one of the exported files
+                shutil.copy("../CMakeLists.txt", ".")
+
+            def build(self):
+                if "MISTAKE" in load(os.path.join(self.source_folder, "CMakeLists.txt")):
+                    raise Exception("MISTAKE BUILD!")
+        """)
+
+    client = TestClient()
+    client.save({"conanfile.py": conanfile,
+                 "conandata.yml": "",
+                 "CMakeLists.txt": "My better cmake"})
+    client.run("install .")
+    client.run("source .")
+    client.run("build .")
+    assert "conanfile.py (mypkg/1.0): Calling build()" in client.out
+
+    # of course create should work too
+    client.run("create .")
+    assert "mypkg/1.0: Created package" in client.out
+
+
+def test_third_party_git_overwrite_build_file():
+    """ Same as the above, but using git clone
+    The trick: "git clone <url> ." needs an empty directory. No reason why the ``src`` folder should
+    be polluted automatically with exports, so just removing things works
+    """
+    git_repo = temp_folder().replace("\\", "/")
+    create_local_git_repo({"CMakeLists.txt": "MISTAKE Cmake"}, folder=git_repo)
+
+    conanfile = textwrap.dedent(r"""
+        import os, shutil
+        from conans import ConanFile, load
+        from conans.tools import save
+
+        class Pkg(ConanFile):
+            name = "mypkg"
+            version = "1.0"
+            exports_sources = "CMakeLists.txt"
+
+            def layout(self):
+                self.folders.source = "src"
+                self.folders.build = "build"
+
+            def source(self):
+                # git clone: WILL FAIL because folder not empty, because UGLY local exports already
+                [shutil.rmtree(t) if os.path.isdir(t) else os.remove(t) for t in os.listdir(".")]
```

Comment:
But I don't know why do you need here to remove anything if you didn't export anything to the src folder.

Reply:
I have added a check to try to explain this, see next commit. The thing is that local ``conan source`` does a copy of CMakeLists.txt and conandata.yml inside the "src" folder.